### PR TITLE
Fix NonZero eager impl.

### PR DESCRIPTION
**Pull request by WilBrady**
**Description**: Nonzero is not a typical unary op as the output shape is not the same as the input self tensor.

**Motivation and Context**
The current nonzero implementation in eager mode is not correctly aligned with aten behavior. First Onnx produces an output that is transposed to that of aten and pytorch. Second the output is a list of indexes and not the same shape as the input tensor. For those reasons, we need to directly implement both nonzero and nonzero.out.

### CODE REVIEW

File: orttraining/orttraining/eager/ort_aten.cpp [hunk 1]
```diff
@@ -957,6 +957,56 @@ at::Tensor& fill__Scalar(
...
+
+  auto status = invoker.Invoke("NonZero", {
+    std::move(ort_input_self),
+  }, ort_outputs_0_NonZero, nullptr);
+
+  if (!status.IsOK())
+    throw std::runtime_error(
+      "ORT return failure status:" + status.ErrorMessage());
+
+  // resize the output and then create output ort value to be updated.
```
Comment (WilBrady) [inline orttraining/orttraining/eager/ort_aten.cpp:987]: need to update comment about calculating transpose shape.

File: orttraining/orttraining/eager/test/ort_ops.py [hunk 2]
```diff
@@ -359,6 +359,20 @@ def test_fill(self):
                 assert cpu_tensor.dtype == ort_tensor.dtype
                 assert torch.equal(cpu_tensor, ort_tensor.to("cpu"))
 
+    def test_nonzero_out(self):
```
[the remark this answers is not in the document]
Comment (WilBrady) [reply orttraining/orttraining/eager/test/ort_ops.py:362]: i test both here. I'll update the test name and add a comment

File: orttraining/orttraining/eager/test/ort_ops.py [hunk 3]
```diff
@@ -359,6 +359,20 @@ def test_fill(self):
...
+        device = self.get_device()
+        cpu_tensor = torch.tensor([[[-1, 0, 1], [0, 1, 0]], [[0, 1, 0], [-1, 0, 1]]])
+        ort_tensor = cpu_tensor.to(device)
```
[the remark this answers is not in the document]
Comment (WilBrady) [reply orttraining/orttraining/eager/test/ort_ops.py:365]: good call, i'll make a change.

File: orttraining/orttraining/eager/test/ort_ops.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (WilBrady) [reply orttraining/orttraining/eager/test/ort_ops.py:362]: I do not test both here but I will

**Review by msftlincoln — APPROVED** [review]

**Review by msftlincoln — APPROVED** [review]

**Review by msftlincoln — APPROVED** [review]

**Pull request merged**